src/parser.c

```diff
@@ -1083,7 +1103,7 @@ AST *parse(span source, const char *filename) {
   p.source = source;
   p.filename = filename;
   p.curr = source.data;
-  p.end = source.data + source.size;
+  p.end = source.data + source.size - 1;
```

Comment:
This seems very wrong. `curr`/`end` is an iterator pair, and as a convention, the `end` iterator points *past* the last element. At least the lexer is written on that assumption, and I’d I could see things going horribly wrong if the last character of a file is not whitespace or in a comment.

I’m honestly not sure why you were having problems w/ `0` bytes in `next_char()`. I’ve written a lot of parsers and I can’t recall that ever happening to me, but perhaps I made some sort of stupid error this time around. Either way, `end` *should* be `data + size` and that *should* just work because it works just fine in literally every other parser I’ve ever written, so I’ll take a look later as to why this seems to not be working in this case...